File: holoken-core/src/main/java/com/holokenmod/creation/GridRandomizer.java
package com.holokenmod.creation;

import androidx.annotation.NonNull;

import com.holokenmod.grid.Grid;
import com.holokenmod.grid.GridCell;
import com.holokenmod.RandomSingleton;

import java.util.ArrayList;
import java.util.Collections;

class GridRandomizer {
	private enum FillMode {
		HORIZONTAL,
		VERTICAL
	}
	
	private final Grid grid;
	private final FillMode fillMode;
	
	GridRandomizer(Grid grid) {
		this.grid = grid;
		
		if (grid.getGridSize().getHeight() > grid.getGridSize().getWidth()) {
			this.fillMode = FillMode.VERTICAL;
		} else {
			this.fillMode = FillMode.HORIZONTAL;
		}
	}
	
	void createGrid() {
		createCells(0, 0);
	}
	
	private boolean createCells(int column, int row) {
		if (column == grid.getGridSize().getWidth()
			|| row == grid.getGridSize().getHeight()) {
			return true;
		}
		
		GridCell cell = grid.getCellAt(row, column);
		
		ArrayList<Integer> possibleDigits;
		
		possibleDigits = getShuffledPossibleDigits(column + row * grid.getGridSize().getWidth());
		
		for(int digit : possibleDigits) {
			cell.setValue(digit);
			
			int nextRow = row;
			int nextColumn = column;
			
			if (fillMode == FillMode.HORIZONTAL) {
				nextColumn++;
				
				if (nextColumn == grid.getGridSize().getWidth()) {
					nextColumn = 0;
					nextRow++;
				}
			} else {
				nextRow++;
				
				if (nextRow == grid.getGridSize().getHeight()) {
					nextRow = 0;
					nextColumn++;
				}
			}
			
			if (createCells(nextColumn, nextRow)) {
				return true;
			}
		}
		
		cell.setValue(GridCell.NO_VALUE_SET);
		
		return false;
	}
	
	@NonNull
	private ArrayList<Integer> getShuffledPossibleDigits(int cellNumber) {
		ArrayList<Integer> possibleDigits;
		
		if (cellNumber == 0) {
			possibleDigits = new ArrayList<>(grid.getPossibleDigits());
		} else {
			possibleDigits = new ArrayList<>();
			
			for (int digit : grid.getPossibleDigits()) {
				if (!grid.isValueUsedInSameRow(cellNumber, digit)
						&& !grid.isValueUsedInSameColumn(cellNumber, digit)) {
					possibleDigits.add(digit);
				}
			}
		}
		
		if (!possibleDigits.isEmpty()) {
			Collections.shuffle(possibleDigits, RandomSingleton.getInstance().getRandom());
		}
		
		return possibleDigits;
	}
}

File: gauguin-core/src/main/kotlin/org/piepmeyer/gauguin/creation/GridRandomizer.kt
package org.piepmeyer.gauguin.creation

import org.piepmeyer.gauguin.grid.Grid
import org.piepmeyer.gauguin.grid.GridCell

class GridRandomizer(
    private val shuffler: PossibleDigitsShuffler,
    private val grid: Grid,
) {
    private enum class FillMode {
        HORIZONTAL,
        VERTICAL,
    }

    private var fillMode =
        if (grid.gridSize.height > grid.gridSize.width) {
            FillMode.VERTICAL
        } else {
            FillMode.HORIZONTAL
        }

    fun createGridValues() {
        createCells(0, 0)
    }

    private fun createCells(
        column: Int,
        row: Int,
    ): Boolean {
        if (column == grid.gridSize.width ||
            row == grid.gridSize.height
        ) {
            return true
        }
        val cell = grid.getValidCellAt(row, column)
        val possibleDigits =
            getShuffledPossibleDigits(
                grid,
                column + row * grid.gridSize.width,
            )

        for (digit in possibleDigits) {
            cell.value = digit
            var nextRow = row
            var nextColumn = column
            if (fillMode == FillMode.HORIZONTAL) {
                nextColumn++
                if (nextColumn == grid.gridSize.width) {
                    nextColumn = 0
                    nextRow++
                }
            } else {
                nextRow++
                if (nextRow == grid.gridSize.height) {
                    nextRow = 0
                    nextColumn++
                }
            }
            if (createCells(nextColumn, nextRow)) {
                return true
            }
        }
        cell.value = GridCell.NO_VALUE_SET
        return false
    }

    private fun getShuffledPossibleDigits(
        grid: Grid,
        cellNumber: Int,
    ): List<Int> {
        val possibleDigits =
            if (cellNumber == 0) {
                grid.variant.possibleDigits
            } else {
                grid.variant.possibleDigits
                    .filter {
                        !grid.isValueUsedInSameRow(cellNumber, it) &&
                            !grid.isValueUsedInSameColumn(cellNumber, it)
                    }.toSet()
            }

        return if (possibleDigits.isEmpty()) {
            emptyList()
        } else {
            shuffler.shufflePossibleDigits(possibleDigits)
        }
    }
}
